### PR TITLE
letsencrypt: Automatically obtain and revoke SSL certificates

### PR DESCRIPTION
**Pull request by JosephKiranBabu**
Let's Encrypt module listens to the following django signals and takes the
appropriate actions.
- domain_added
- domain_removed
- domainname_change

Signed-off-by: Joseph Nuthalpati <njoseph@thoughtworks.com>

### CODE REVIEW

**Comment by JosephKiranBabu** [comment]
closes #935

**Comment by JoKeyser** [comment]
To whom considering reviewing this (I'd be willing to), note that this will clash (temporarily) with PR #919. So maybe consider reviewing this first...

**Comment by JosephKiranBabu** [comment]
Fixed conflicts.
Travis build is failing because configobj package is not available in the build environment.

**Comment by jvalleroy** [comment]
Sorry for the late review. pyflakes found some issues:
```
$ pyflakes3 plinth/modules/letsencrypt/__init__.py 
plinth/modules/letsencrypt/__init__.py:171: undefined name 'domain'
plinth/modules/letsencrypt/__init__.py:188: undefined name 'domain'

$ pyflakes3 plinth/modules/letsencrypt/views.py plinth/modules/letsencrypt/views.py:33: 'plinth.modules.names' imported but unused
plinth/modules/letsencrypt/views.py:86: undefined name 'config'
```

**Comment by JosephKiranBabu** [comment]
Fixed conflicts and failing tests. Ready for review.

File: plinth/modules/letsencrypt/__init__.py [hunk 1]
```diff
@@ -125,3 +148,65 @@ def get_installed_modules():
...
+        if domain_name == name and \
+           domain_status.certificate_available and \
+           domain_status.validity == 'valid':
+            return False
+
+    try:
+        # Obtaining certs during tests isn't expected to succeed
+        if sender != 'test':
+            try_action(name, 'obtain')
+            enable_renewal_management(domain)
```
Comment (jvalleroy) [inline plinth/modules/letsencrypt/__init__.py]: What does ```domain``` refer to here?

File: plinth/modules/letsencrypt/__init__.py [hunk 2]
```diff
@@ -125,3 +148,65 @@ def get_installed_modules():
...
+    """Revoke Let's Encrypt certificate for the removed domain"""
+    try:
+        if sender != 'test':
+            try_action(name, 'revoke')
+        logger.info("Revoked the Let\'s Encrypt certificate for " + name)
+        return True
+    except ActionError as exception:
+        logger.warn(
+            ('Failed to revoke certificate for domain {domain}: {error}')
+            .format(domain=domain, error=exception.args[2]))
```
Comment (jvalleroy) [inline plinth/modules/letsencrypt/__init__.py]: Also here, where is ```domain``` defined?

File: plinth/modules/letsencrypt/views.py [hunk 3]
```diff
@@ -71,19 +69,22 @@ def revoke(request, domain):
...
             request,
             _('Failed to obtain certificate for domain {domain}: {error}')
             .format(domain=domain, error=exception.args[2]))
-        successful_obtain = False
+    return redirect(reverse_lazy('letsencrypt:index'))
+
 
-    if domain == config.get_domainname() and successful_obtain:
+def enable_renewal_management(request, domain):
+    if domain == config.get_domainname():
```
Comment (jvalleroy) [inline plinth/modules/letsencrypt/views.py]: ```config``` is still used here, but no longer imported.

**Comment by JosephKiranBabu** [comment]
@jvalleroy Looks like I had flake8 disabled in my Emacs all along. Fixed that. Will update this PR soon.

**Comment by jvalleroy** [comment]
I noticed these logs when starting plinth (2 separate restarts):
```
[2017-10-12 02:43:19,912] plinth.actions INFO     Executing command - ['sudo', '-n', '/usr/share/plinth/actions/letsencrypt', 'revoke', '--domain', '']
[2017-10-12 02:43:21,144] plinth.actions ERROR    Error executing command - ['sudo', '-n', '/usr/share/plinth/actions/letsencrypt', 'revoke', '--domain', ''], , Requested domain  is not a FQDN because it contains an empty label.


[2017-10-12 02:43:21,144] plinth.modules.letsencrypt WARNING  Failed to revoke certificate for domain : Requested domain  is not a FQDN because it contains an empty label.
```
But after the 3rd restart, I didn't see them again.

I have a domain name set, so no idea why it is trying to revoke an empty domain.

**Comment by jvalleroy** [comment]
For testing, I removed my current domain name from Configure. When I click the Certificates page, I get this error:
```
django.urls.exceptions.NoReverseMatch: Reverse for 'toggle_hooks' with arguments '('',)' not found. 1 pattern(s) tried: ['sys/letsencrypt/toggle_hooks/(?P<domain>[^/]+)/$']
```
Which looks like the same error as in #1040.

**Comment by JoKeyser** [comment]
@jvalleroy this might indeed be a bug of my implementation. Sorry I won't have time soon to fix this, unfortunately. @JosephKiranBabu would be great if you could fix this in this PR?

**Comment by JosephKiranBabu** [comment]
When a new domain name is assigned, the old domain name is taken as an empty string and is tried to be revoked.
I'm really puzzled on how this is getting triggered, though I can simply apply a band-aid in the revoke method to check whether the domain name is not empty.

**Comment by JosephKiranBabu** [comment]
The travis build is failing though we've added configobj to the `requirements.txt` file.

**Comment by jvalleroy** [comment]
Merged along with #1073.

**Pull request closed without merging**